src/Mod/Sketcher/Gui/DrawSketchHandlerBSplineByInterpolation.h

```diff
@@ -0,0 +1,606 @@
+                text.sprintf(" (%s, %s)", lengthString.c_str(), angleString.c_str());
+                setPositionText(position, text);
+            }
+        }
+    }
+
+    void eraseEditCurve()
+    {
+        drawEdit(std::vector<Base::Vector2d>());
+    }
+
+    bool finishCommand(Base::Vector2d position)
+    {
+        if (Mode==STATUS_CLOSE) {
+            // TODO: make modifications for construction by knots
+            unsetCursor();
+            resetPositionText();
+
+            unsigned int myDegree = 3;
+
+            if (ConstrMethod == 0) {
+                BSplineMults.front() = myDegree + 1; // FIXME: This is hardcoded until degree can be changed
+                BSplineMults.back() = myDegree + 1; // FIXME: This is hardcoded until degree can be changed
+            }
+
+            std::vector<std::stringstream> streams;
+
+            // The first point
+            streams.emplace_back();
+            streams.back() << "App.Vector(" << BSplineKnots.front().x << "," << BSplineKnots.front().y << "),";
+            // Middle points
+            for (size_t i = 1; i < BSplineKnots.size() - 1; ++i) {
+                streams.back() << "App.Vector(" << BSplineKnots[i].x << "," << BSplineKnots[i].y << "),";
+                if (BSplineMults[i] >= myDegree) {
+                    streams.emplace_back();
+                    streams.back() << "App.Vector(" << BSplineKnots[i].x << "," << BSplineKnots[i].y << "),";
+                }
+            }
+            // The last point
+            streams.back() << "App.Vector(" << BSplineKnots.back().x << "," << BSplineKnots.back().y << "),";
+
+            std::vector<std::string> controlpointses; // Gollum gollum
+            controlpointses.reserve(streams.size());
+            for (auto & stream: streams) {
+                // TODO: Use subset of points between C0 knots.
+                controlpointses.emplace_back(stream.str());
+

+                auto & controlpoints = controlpointses.back();
+
+                // remove last comma and add brackets
+                int index = controlpoints.rfind(',');
+                controlpoints.resize(index);
+
+                controlpoints.insert(0,1,'[');
+                controlpoints.append(1,']');
+            }
+
+            // With just 3 points provided OCCT gives a quadratic spline where
+            // the middle point is NOT a knot. This needs to be treated differently.
+            // FIXME: Decide whether to force a knot or not.
+            std::vector<bool> isBetweenC0Points(BSplineKnots.size(), false);
+            for (size_t i = 1; i < BSplineKnots.size()-1; ++i) {
+                if (BSplineMults[i-1] >= myDegree && BSplineMults[i+1] >= myDegree)
+                    isBetweenC0Points[i] = true;
+            }
+
+            int currentgeoid = getHighestCurveIndex();
+
+            try {
+                //Gui::Command::openCommand(QT_TRANSLATE_NOOP("Command", "Add B-spline curve"));

```

Comment:
Leftover?

Reply:
Leftover. Though in the undo history it reads as "create knot" or something, like for the vanilla tool. Should do something about that.

Reply:
I have checked that undo works in my tests, but I did not look into the string itself. It is always good to be expressive in the naming of the undo history, but be practical. 

If the problem you are facing is that you have created a command for the knot that then you would like to include the B-Spline creation in it, perhaps we should look into extending the framework to enable renaming of an open not yet committed transaction (it may even be possible and me not knowing it).

Reply:
The problem is that if once the spline is added, the undo history still only says "add a knot", instead of "add a B-spline". Ideally, when editing, undo should remove knots, but when editing is done, undo should remove the whole B-spline.